cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ZookeeperLeaderFinder.java

```diff
@@ -0,0 +1,267 @@
+import io.vertx.core.net.NetSocket;
+import io.vertx.core.net.PemKeyCertOptions;
+import io.vertx.core.net.PemTrustOptions;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.security.cert.CertificateFactory;
+import java.util.Base64;
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Helper class for finding the leader of a ZK cluster
+ */
+public class ZookeeperLeaderFinder {
+
+    private static final Logger log = LogManager.getLogger(ZookeeperLeaderFinder.class);
+
+    static final int UNKNOWN_LEADER = -1;
+
+    private final Vertx vertx;
+    private final SecretOperator secretOperator;
+    private final Supplier<BackOff> backOffSupplier;
+
+    public ZookeeperLeaderFinder(Vertx vertx, SecretOperator secretOperator, Supplier<BackOff> backOffSupplier) {
+        this.vertx = vertx;
+        this.secretOperator = secretOperator;
+        this.backOffSupplier = backOffSupplier;
+    }
+
+    /*test*/ NetClientOptions clientOptions(CertAndKey coCertKey, Secret clusterCaCertificateSecret)
+            throws GeneralSecurityException, IOException {
+
+        CertificateFactory x509 = CertificateFactory.getInstance("X.509");
+        Base64.Decoder decoder = Base64.getDecoder();
+
+        // Validate the certificates are not corrupt
+        //x509.generateCertificate(new ByteArrayInputStream(decoder.decode(clusterCaCertificateSecret.getData().get(Ca.CA_CRT))));
+        //x509.generateCertificate(new ByteArrayInputStream(coCertKey.cert()));
+
+        Buffer coCertBuffer = Buffer.buffer(coCertKey.cert());
+        log.debug("CO cert\n{}", coCertBuffer);
+        Buffer coKeyBuffer = Buffer.buffer(coCertKey.key());
+        //log.debug("CO key\n{}", coKeyBuffer);
+        Buffer clusterCaCertBuffer = Buffer.buffer(decoder.decode(clusterCaCertificateSecret.getData().get(Ca.CA_CRT)));
+        log.debug("Cluster CA cert\n{}", clusterCaCertBuffer);
+        return new NetClientOptions()
+                .setConnectTimeout(10_000)
+                .setSsl(true)
+                //.setHostnameVerificationAlgorithm("HTTPS")
+                .setPemKeyCertOptions(new PemKeyCertOptions()
+                        .setCertValue(coCertBuffer)
+                        .setKeyValue(coKeyBuffer))
+                .setPemTrustOptions(new PemTrustOptions()
+                        .addCertValue(clusterCaCertBuffer));
+    }
+
+    /**
+     * Returns a Future which completes with the the id of the Zookeeper leader.
+     * An exponential backoff is used if no ZK node is leader on the attempt to find it.
+     * If there is no leader after 3 attempts then the returned Future completes with {@link #UNKNOWN_LEADER}.
+     */
+    Future<Integer> findZookeeperLeader(String cluster, String namespace, List<Pod> pods) {
+        if (pods.size() <= 1) {
```

Comment:
In that case we're going to end up trying to roll the pod (whether it was created or not) irrespective whether we return 0 or -1 (=`UNKNOWN_LEADER`), so what difference does it make? If the operator is going to cope with the non-creation of Pods I think it would have to do it at a higher level than here.

Reply:
Well, that is why I asked ... because I was not sure whether it would or wouldn't make a difference ;-). If it makes no difference, I'm fine with this.